File: engine/galaxy.py
import random

class GalaxyUniverse:
    """
    Represents the Primus Universe structure — orbits, planets, and moons.
    Supports trust-biased recursive traversal.
    """

    def __init__(self, map_data):
        self.map = map_data or {}
        self.orbits = self.map.get("orbits", {})
        self.planets = self.map.get("planets", {})
        self.moons = self.map.get("moons", {})

    def random_path(self, trustmap=None):
        """
        Selects an orbit → planet → moon path.
        If a trustmap is provided, selection is biased by trust values.
        """

        if not self.planets:
            return "Void", "NoPlanet", "NoMoon"

        trustmap = trustmap or {}

        # Build weighted candidates
        candidates = []
        for planet in self.planets:
            orbit = self.orbits.get(planet, "UnknownOrbit")
            moon_list = self.moons.get(planet, ["NoMoon"])

            for moon in moon_list:
                key = f"{orbit}:{planet}:{moon}"
                weight = trustmap.get(key, 1)  # default trust weight is 1
                candidates.append((orbit, planet, moon, weight))

        if not candidates:
            return "Void", "NoPlanet", "NoMoon"

        # Weighted random choice
        total = sum(w for _, _, _, w in candidates)
        pick = random.uniform(0, total)
        cumulative = 0

        for orbit, planet, moon, weight in candidates:
            cumulative += weight
            if pick <= cumulative:
                return orbit, planet, moon

        # Fallback
        return random.choice([(o, p, m) for o, p, m, _ in candidates])

File: engine/galaxy.ts
// File: engine/galaxy.ts

export class GalaxyUniverse {
  orbits: Record<string, string>;
  planets: Record<string, string[]>;

  constructor(genesis: { orbits: Record<string, string>; planets: Record<string, string[]> }) {
    this.orbits = genesis.orbits;
    this.planets = genesis.planets;
  }

  random_orbit(): string {
    const keys = Object.values(this.orbits);
    return keys[Math.floor(Math.random() * keys.length)];
  }

  random_planet(orbit: string): string {
    const matching = Object.entries(this.orbits)
      .filter(([_, o]) => o === orbit)
      .map(([p]) => p);
    return matching[Math.floor(Math.random() * matching.length)];
  }

  random_moon(planet: string): string {
    const moons = this.planets[planet] || [];
    if (moons.length === 0) return "none";
    return moons[Math.floor(Math.random() * moons.length)];
  }
}
